Add support for Nessie Catalog in Iceberg Connector

nastra opened the pull request:
This PR integrates the [Nessie catalog functionality](https://github.com/apache/iceberg/tree/master/nessie/src/main/java/org/apache/iceberg/nessie) to the Iceberg connector.
Roughtly, it adds the following new things:

* adds a new `NESSIE` catalog type, that uses `NessieCatalog`
* renames `IcebergHadoopMetadata` to `IcebergNativeMetadata` and adds the particular `CatalogType`. This allows reusing that class for both `HADOOP` + `NESSIE`.
* adds 2 new `IcebergSessionProperties` to control which referenceName / referenceHash to use with Nessie
* some minor adjustments in `IcebergResourceFactory` that make sure to pass the correct properties to the `NessieCatalog`
* adds `NessieConfig` that controls different Nessie settings and passes those to the underyling `NessieCatalog`
* adds `NessieContainer` that allows to bring up a Nessie server for testing
* additional details about Nessie itself can be found at https://projectnessie.org/

```
== RELEASE NOTES ==

Iceberg Changes
* Add support for NessieCatalog
```

nastra commented:
@ChunxuTang could you review this please?

ChunxuTang commented:
@nastra Thanks for your work! Sure, happy to help review it soon this week.

nastra commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestNessieMultiBranching.java:
```diff
@@ -0,0 +1,169 @@
+        assertQuerySucceeds(sessionTwo, "CREATE SCHEMA namespace_two");
+        assertQuerySucceeds(sessionTwo, "SHOW SCHEMAS FROM iceberg LIKE 'namespace_two'");
+
+        // TODO: enable this after bump to Iceberg 0.14.0
```
this requires https://github.com/apache/iceberg/pull/4700 to work as expected

nastra commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestNessieMultiBranching.java:
```diff
@@ -0,0 +1,169 @@
+        Session sessionOne = sessionOnRef(one.getName());
+        Session sessionTwo = sessionOnRef(two.getName());
+        assertQuerySucceeds(sessionOne, "CREATE SCHEMA namespace_one");
+        assertQuerySucceeds(sessionOne, "SHOW SCHEMAS FROM iceberg LIKE 'namespace_one'");
```
this and the one below will be changed to `assertThat(computeActual(sessionOne, "SHOW SCHEMAS FROM iceberg LIKE 'namespace_one'").getMaterializedRows()).hasSize(1);` once Iceberg 0.14.0 is out, since it requires https://github.com/apache/iceberg/pull/4700 to properly work

ChunxuTang commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergDistributedNessie.java:
```diff
@@ -0,0 +1,64 @@
+    @AfterClass
+    public void tearDown()
+    {
+        if (null != nessieContainer) {
```
nit: I think in the presto codebase we mainly use `xx != null` instead of `null != xx`. In this PR, there're some places where you use `xx != null` but some `null != xx`. Could you make them consistent?

nastra replied:
sure, fixed

ChunxuTang commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSmokeNessie.java:
```diff
@@ -0,0 +1,62 @@
+    @AfterClass
+    public void tearDown()
+    {
+        if (null != nessieContainer) {
```
nit: != null

nastra replied:
fixed

ChunxuTang commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSystemTablesNessie.java:
```diff
@@ -0,0 +1,82 @@
+    public void tearDown()
+    {
+        super.tearDown();
+        if (null != nessieContainer) {
```
nit: != null

nastra replied:
fixed

ChunxuTang commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/nessie/TestIcebergSystemTablesNessie.java:
```diff
@@ -0,0 +1,82 @@
+                .put("hive.metastore.catalog.dir", catalogDir.toFile().toURI().toString())
+                .put("iceberg.catalog.type", NESSIE.name())
+                .put("iceberg.nessie.uri", nessieContainer.getRestApiUri())
+                .put("iceberg.catalog.warehouse", catalogDir.getParent().toFile().toURI().toString())
```
Just curious: is `iceberg.catalog.warehouse` required for the Nessie catalog?

nastra replied:
yes this is in fact required

ChunxuTang reviewed (commented):
@nastra Thanks so much for your nice contribution! This is a very helpful feature!
Just left some minor suggestions. Feel free to take a look.

ChunxuTang reviewed (approved):
LGTM! @nastra Thanks again for your nice work!

Hi @beinan and @kewang1024, would you also like to take a review of this PR? Thanks!

beinan commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java:
```diff
@@ -134,6 +144,29 @@ public Map<String, String> getCatalogProperties(ConnectorSession session)
+            nessieConfig.getConnectTimeoutMillis().ifPresent(val -> properties.put("transport.connect-timeout", val.toString()));
+            nessieConfig.getClientBuilderImpl().ifPresent(val -> properties.put("client-builder-impl", val));
+            nessieConfig.getAuthenticationType().ifPresent(type -> {
+                if (type.equals(BASIC) && nessieConfig.getUsername().isPresent() && nessieConfig.getPassword().isPresent()) {
```
nit: how about throwing an exception when type is basic but username or password is not provided?   your call

nastra replied:
makes sense, updated the code

beinan commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java:
```diff
@@ -134,6 +144,29 @@ public Map<String, String> getCatalogProperties(ConnectorSession session)
+                    properties.put("authentication.token", nessieConfig.getBearerToken().get());
+                }
+            });
+            if (!nessieConfig.isCompressionEnabled()) {
```
nit: why not make this config as isCompressionDisabled?

nastra replied:
makes sense, fixed

beinan reviewed (commented):
lgtm except a couple of very minor issues.

nastra commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java:
```diff
@@ -66,7 +72,11 @@ public IcebergResourceFactory(IcebergConfig config, IcebergCatalogName catalogNa
-            return catalogCache.get(getCatalogCacheKey(session), () -> CatalogUtil.loadCatalog(
+            String catalogCacheKey = getCatalogCacheKey(session);
+            if (catalogType == NESSIE) {
+                catalogCacheKey = catalogCacheKey + IcebergSessionProperties.getNessieReferenceName(session) + "@" + IcebergSessionProperties.getNessieReferenceHash(session);
```
done

kewang1024 commented:
Hey sorry for the delay, will finish review by the end of today (5/17)

kewang1024 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSessionProperties.java:
```diff
@@ -240,6 +244,16 @@ public IcebergSessionProperties(
+                        false),
+                stringProperty(
+                        NESSIE_REFERENCE_HASH,
+                        "Nessie reference hash to use",
```
Why don't we have config for NESSIE_REFERENCE_HASH?

nastra replied:
because this is generally something a user would only specify at runtime when actually reading data at a particular hash. Having a config option for that would just be confusing to users because they would have to know the hash upfront.

kewang1024 replied:
So does it mean usually different users would use different hash and they don't usually share? and thus we don't have any need to set a default value on a cluster level

kewang1024 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/nessie/NessieConfig.java:
```diff
@@ -0,0 +1,171 @@
+    @ConfigDescription("The default Nessie reference to work on")
+    public NessieConfig setDefaultReferenceName(String defaultReferenceName)
+    {
+        this.defaultReferenceName = defaultReferenceName;
```
Should we have isNullOrEmpty check here?

nastra replied:
I don't see this kind of validation being done in other Config classes, so I was assuming that the `@NotNull` annotation on `getDefaultReferenceName()` would take care of this.

kewang1024 replied:
I think we do have null check examples, `StaticCatalogStoreConfig`

nastra replied:
added a null/empty check + added a test for it

nastra replied:
Just for compleness: I went over the config code to understand when things are being validated. I used the below test and temporarily removed the null/empty check in `setDefaultReferenceName`. Note that we provide an empty string to `iceberg.nessie.ref` in this example. 

```
@Test
public void testNessieMetastore()
{
    getOnlyElement(new IcebergPlugin().getConnectorFactories()).create(
                    "test",
                    ImmutableMap.of(
                            "hive.metastore.uri", "thrift://foo:1234",
                            "iceberg.catalog.type", "nessie",
                            "iceberg.nessie.ref", ""),
                    new TestingConnectorContext())
            .shutdown();
}
```

The validation of the config basically happens when the Plugin is being bootstrapped [here](https://github.com/prestodb/presto/blob/99748b79827022e9633103ee8948cd48b466a241/presto-iceberg/src/main/java/com/facebook/presto/iceberg/InternalIcebergConnectorFactory.java#L83-L86) and this is when the validation annotations (`@NotNull` / `@NotEmpty`/...) are being checked/verified. The startup of the plugin then correctly failed with the following exception:

```
com.google.inject.CreationException: Unable to create injector, see the following errors:

1) Error: Invalid configuration property iceberg.nessie.ref: must not be null or empty (for class com.facebook.presto.iceberg.nessie.NessieConfig.defaultReferenceName)

1 error

	at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:543)
	at com.google.inject.internal.InternalInjectorCreator.initializeStatically(InternalInjectorCreator.java:159)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:106)
	at com.google.inject.Guice.createInjector(Guice.java:87)
	at com.facebook.airlift.bootstrap.Bootstrap.initialize(Bootstrap.java:257)
	at com.facebook.presto.iceberg.InternalIcebergConnectorFactory.createConnector(InternalIcebergConnectorFactory.java:86)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.facebook.presto.iceberg.IcebergConnectorFactory.create(IcebergConnectorFactory.java:49)
	at com.facebook.presto.iceberg.TestIcebergPlugin.testNessieMetastore(TestIcebergPlugin.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
```
This would explain why the config classes don't perform validation checks in the setters but completely rely on the Validation annotations on the getters.

kewang1024 replied:
Thanks for the fact check on the config validation!! Then looks like the null check could actually be safely removed since the getter would perform the null check?

nastra replied:
yep I removed the null check now

kewang1024 reviewed (commented):
For all the four tests files, they share the same init/teardown/createQueryRunner function, can we extract those out to a common place so we don't need to create repeated logic

nastra commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -105,6 +105,58 @@ The path(s) for Hadoop configuration resources. Example:
+                                                     Example: ``5000``
+
+``iceberg.nessie.connect-timeout-ms``                The connection timeout in milliseconds for connection
+                                                     connection requests to the Nessie server.
```
good catch, fixed

nastra commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -105,6 +105,58 @@ The path(s) for Hadoop configuration resources. Example:
+                                                     connection requests to the Nessie server.
+                                                     Example: ``10000``
+
+``iceberg.nessie.compression-enabled``               Configure whether compression should be enabled or not for
```
fixed

nastra commented on presto-docs/src/main/sphinx/connector/iceberg.rst:
```diff
@@ -105,6 +105,58 @@ The path(s) for Hadoop configuration resources. Example:
+``iceberg.nessie.compression-enabled``               Configure whether compression should be enabled or not for
+                                                     requests to the Nessie server, defaults to ``true``.
+
+``iceberg.nessie.client-builder-impl``               Configure the custom ClientBuilder implementation class
```
fixed

nastra commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java:
```diff
@@ -66,7 +74,11 @@ public IcebergResourceFactory(IcebergConfig config, IcebergCatalogName catalogNa
         try {
-            return catalogCache.get(getCatalogCacheKey(session), () -> CatalogUtil.loadCatalog(
+            String catalogCacheKey = getCatalogCacheKey(session);
+            if (catalogType == NESSIE) {
+                catalogCacheKey = catalogCacheKey + getNessieReferenceName(session) + "@" + getNessieReferenceHash(session);
+            }
```
indeed that makes sense, fixed

nastra commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java:
```diff
@@ -134,6 +146,32 @@ public Map<String, String> getCatalogProperties(ConnectorSession session)
+            nessieConfig.getConnectTimeoutMillis().ifPresent(val -> properties.put("transport.connect-timeout", val.toString()));
+            nessieConfig.getClientBuilderImpl().ifPresent(val -> properties.put("client-builder-impl", val));
+            nessieConfig.getAuthenticationType().ifPresent(type -> {
+                if (type.equals(BASIC)) {
```
fixed

nastra commented:
> For all the four tests files, they share the same init/teardown/createQueryRunner function, can we extract those out to a common place so we don't need to create repeated logic

I moved out the required connector properties for Nessie into a util method but I'm not sure it's worth extracting the init/teardown to a common place, since we're essentially really only calling start/stop on the container. There's currently no other complex init/teardown required for Nessie so imo it seems (at least for now) to be better to leave the start/stop of the container in the respective test classes.

nastra commented:
Thanks for your reviews @kewang1024 @beinan @ChunxuTang. I rebased the PR and I hope I addressed everything that's relevant. Please take another look if you can.

kewang1024 commented:
> > For all the four tests files, they share the same init/teardown/createQueryRunner function, can we extract those out to a common place so we don't need to create repeated logic
> 
> I moved out the required connector properties for Nessie into a util method but I'm not sure it's worth extracting the init/teardown to a common place, since we're essentially really only calling start/stop on the container. There's currently no other complex init/teardown required for Nessie so imo it seems (at least for now) to be better to leave the start/stop of the container in the respective test classes.

So essentially what I meant was creating a parent class that has init/teardown/createQueryRunner function for nessie and then all your four tests files can only focus on its own test logic

beinan commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergMetadataFactory.java:
```diff
@@ -56,7 +56,8 @@ public ConnectorMetadata create()
         switch (catalogType) {
             case HADOOP:
-                return new IcebergHadoopMetadata(resourceFactory, typeManager, commitTaskCodec);
+            case NESSIE:
```
Nit: to be more consistent, we could put all the cases in alphabetical order

nastra replied:
that's what I had initially, but Intellij would then complain due to duplicate case branches. So I figured that it's better to merge `HADOOP` + `NESSIE` together

nastra replied:
@beinan let me know if you would like me to still change this and introduce duplicate case branches

beinan reviewed (commented):
lgtm, just one non-blocking minor issue. could please address all the comments from @kewang1024 , then we could merge your PR. Thanks a lot!

nastra commented:
> > > For all the four tests files, they share the same init/teardown/createQueryRunner function, can we extract those out to a common place so we don't need to create repeated logic
> > 
> > 
> > I moved out the required connector properties for Nessie into a util method but I'm not sure it's worth extracting the init/teardown to a common place, since we're essentially really only calling start/stop on the container. There's currently no other complex init/teardown required for Nessie so imo it seems (at least for now) to be better to leave the start/stop of the container in the respective test classes.
> 
> So essentially what I meant was creating a parent class that has init/teardown/createQueryRunner function for nessie and then all your four tests files can only focus on its own test logic

@kewang1024 that would be ideal if we could have a base class, but all of those 4 test classes are already extending different other test classes to test different things so I don't see how we could have a base class for this unfortunately.

beinan reviewed (approved):
lgtm
